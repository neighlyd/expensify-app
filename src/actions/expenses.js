import database from '../firebase/firebase'

// component calls action generator
// action generator returns a function
// component dispatches the function
// Redux middleware to run the function.
// function runs -> has the ability to dispatch other actions and do whatever it wants.

// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const {
            description = '', 
            note = '', 
            amount = 0, 
            createdAt = 0
        } = expenseData
        const expense = { description, note, amount, createdAt }

        const uid = getState().auth.uid

        return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }))
        })
    }
}

// REMOVE_EXPENSE
export const removeExpense = ( {id} ) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = (({id}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/expenses`).child(id).remove().then((ref) => {
            dispatch(removeExpense({id}))
        })
    }
})

// EDIT_EXPENSE

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

export const startEditExpense = ((id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid

        return database.ref(`users/${uid}/expenses`).child(id).update(updates).then((snapshot) =>{
            dispatch(editExpense(id, updates))
        })
    }
})

// SET_EXPENSES - used to set up our expenses array in the redux store once it is retrieved from the firebase data store.
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
})

export const startSetExpenses = () => {
    // Fetch all expense data once
    // Parse data into an array
    // Dispatch SET_EXPENSES with new data array

    return (dispatch, getState) => {
        const uid = getState().auth.uid

        return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
            const expenses = []
            
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                })
            })
            dispatch(setExpenses(expenses))
        })
    }
}