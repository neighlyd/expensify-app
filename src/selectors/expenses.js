import moment from 'moment';

// Get visible expenses
// startDate and endDate are given as moment() instances.
export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day'): true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day'): true;
        const textMatch = expense.description.toLowerCase().includes(text.trim().toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1;
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1;
        }
    });
};