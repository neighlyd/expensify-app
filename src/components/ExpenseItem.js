import React from 'react';
import { connect } from 'react-redux';

import { removeExpense } from '../actions/expenses';


const ExpenseItem = ({ dispatch, id, description, amount, createdAt }) => (
    <div>
        <h3>Description: { description}</h3>
        <p>Amount: ${ amount}</p>
        <p>Created At: { createdAt}</p>
        <button onClick={(e) => dispatch(removeExpense({ id }))}>Remove</button>
        <hr/>
    </div>
);

export default connect()(ExpenseItem);