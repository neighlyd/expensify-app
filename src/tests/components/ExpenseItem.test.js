import React from 'react';
import { shallow } from 'enzyme';

import ExpenseItem from '../../components/ExpenseItem';
import expenses from '../fixtures/expenses';

test('should render expense item', () => {
    const wrapper = shallow(<ExpenseItem {...expenses[0]}/>);
    expect(wrapper).toMatchSnapshot();
})