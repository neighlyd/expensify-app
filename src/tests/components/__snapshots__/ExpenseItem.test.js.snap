// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense item 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      Description: 
      Gum
    </h3>
  </Link>
  <p>
    Amount: $
    1.95
  </p>
  <p>
    Created At: 
    Dec 31st, 1969
  </p>
  <hr />
</div>
`;
